compiler/ir/backend.wasm/src/org/jetbrains/kotlin/backend/wasm/ir2wasm/BodyGenerator.kt

```diff
@@ -0,0 +1,490 @@
+
+                generateTypeRTT(toType)
+                body.buildRefCast(context.transformType(fromType), context.transformBoxedType(toType))
+                generateInstanceFieldAccess(field)
+            }
+            else -> {
+                return false
+            }
+        }
+        return true
+    }
+
+    override fun visitContainerExpression(expression: IrContainerExpression) {
+        val statements = expression.statements
+        if (statements.isEmpty()) return
+
+        statements.dropLast(1).forEach {
+            statementToWasmInstruction(it)
+        }
+
+        if (expression.type != irBuiltIns.unitType) {
+            generateExpression(statements.last() as IrExpression)
+        } else {
+            statementToWasmInstruction(statements.last())
+        }
+    }
+
+    override fun visitBreak(jump: IrBreak) {
+        body.buildBr(context.referenceLoopLevel(jump.loop, LoopLabelType.BREAK))
+    }
+
+    override fun visitContinue(jump: IrContinue) {
+        body.buildBr(context.referenceLoopLevel(jump.loop, LoopLabelType.CONTINUE))
+    }
+
+    override fun visitReturn(expression: IrReturn) {
+        generateExpression(expression.value)
+
+        // FIXME: Hack for "returning" Unit from functions with generic return type.
+        //        Common case -- lambdas returning unit.
+        if (expression.value.type == irBuiltIns.unitType &&
+            expression.returnTargetSymbol.owner.returnType(backendContext) != irBuiltIns.unitType
+        ) {
+            val irReturnType = expression.returnTargetSymbol.owner.returnType(backendContext)
+
+            if (irReturnType != irBuiltIns.unitType) {
+                generateDefaultInitializerForType(context.transformType(irReturnType), body)
+            }
+        }
+
+        body.buildInstr(WasmOp.RETURN)
+    }
+
+    override fun visitWhen(expression: IrWhen) {
+        if (expression.type == irBuiltIns.unitType) {
+            var ifCount = 0
+            for (branch in expression.branches) {
+                if (!isElseBranch(branch)) {
+                    generateExpression(branch.condition)
+                    body.buildIf(label = null, resultType = null)
+                    statementToWasmInstruction(branch.result)
+                    body.buildElse()
+                    ifCount++
+                } else {
+                    statementToWasmInstruction(branch.result)
+                    break
+                }
+            }
+
+            repeat(ifCount) { body.buildEnd() }
+            return
+        }
+
+        val resultType = context.transformBlockResultType(expression.type)
+        var ifCount = 0
+        for (branch in expression.branches) {
+            if (!isElseBranch(branch)) {
+                generateExpression(branch.condition)
+                body.buildIf(null, resultType)
+                generateExpression(branch.result)
+                if (expression.type == irBuiltIns.nothingType) {
```

Comment:
unreachable will be inserted by `generateExpression` in this case

Reply:
`generateExpression` will insert unreachable based on `branch.result.type` which may not be `irBuiltIns.nothingType` when the whole when-expression has type `irBuiltIns.nothingType`

Reply:
thanks, I've missed it

Reply:
couldn't we add one `unreachable` after all ifs?

Reply:
?